docs/PROCEDURES.md

```diff
@@ -0,0 +1,69 @@
+# Additional procedures for code maintenance 
+
+## Managing pinned dependencies
+
+The OpenSSF, via the [scorecard](https://securityscorecards.dev/) project recommends that projects pin any
+dependencies they use:
+
+* to ensure reproducibility
+* to reduce the change for rogue dependency updates to compromise software
+
+It's important to note that this requires any changes to dependencies are properly reviews, and
+these changes, by design, should not be automatic in themselves, though automated tools may provide recommendations.
+
+### Python dependencies
+
+Python dependencies used in the build process should be pinned to a specific version to ensure reproducibility.
+
+This is done by:
+ * Using the `--require-hashes` option on any `pip install` command line
+ * Adding the required hash in the `requirements.txt`
+
+ Currently this is used withou `.github/workflows` but the same principle applies elsewhere.
+
+ To make this easier, a version of the `requirements.txt` without hashes has been saved as `requirements.in`. This is
+ to make maintenance easier, but it is not used at script execution time.
+
+ The `pip-compile` tool must be installed via the [pip-tools](https://pypi.org/project/pip-tools/) package.
+ 
+ To add a new, or changed dependency:
+ ```
+ pip-compile --generate-hashes --output-file=requirements.txt requirements.in
+ ```
+
+ This will update requirements.txt with the correct hashes.
+
+ Correction function should be verified, and then both `requirements.txt` and `requirements.in` checked in.
+
+### Github Actions
+
+All actions used in `.github/worfklows' should pin the exact version of the action they are using, for
+example a step such as:
+
+```yaml
+- name: Checkout code
+  uses: actions/checkout@0ad4b8fadaa221de15dcec353f45205ec38ea70b # pin@v4
+```
+
+The exact hash specified after `@` is the git commit hash within the repo where the action is found.
+
+The [pin github action](https://github.com/mheap/pin-github-action) tool can be used to maintain these
+by, for example, running:
+
+```shell
+pin-github-action unix.yml
+```
+
+This will update the hashes with the latest version if no pinned value is set.
+
+For major updates, update the comment ie `pin@v4` to `pin@v5` and the tool will attempt to find the new hash.
+
+To help in explanation here's an example of a similar code fragment between tool executions:
+
+* Original entry is `uses: actions/checkout@v3`
+* run `pin-github-action unix.yml`
+* We now see `uses: actions/checkout@f43a0e5ff2bd294095638e18286ca9a3d1956744 # pin@v3`
+* later we want to go to v4, so update the text to `uses: actions/checkout@f43a0e5ff2bd294095638e18286ca9a3d1956744 # pin@v4`
+* Now run `pin-github-action unix.yml` to correct the sha
+* File now shows `uses: actions/checkout@692973e3d937129bcbf40652eb9f2f61becf3332 # pin@v4`
```

Comment:
Ah, now I get it: The tool relies on the comments it generated in previous runs, moving the original version string in the code ("v3" in the example) to a comment, so one now has to edit the comments and then run the tool if one wants to upgrade what's actually executed, right?

Reply:
yes, that's right.

We may start with 'actions/myaction@2.1'
Then run the tool and get `actions/myaction@a1b2c3d4 # pin@2.1' (etc)
Then run the tool again and get `actions/myaction@e5f6a1 # pin@2.1' if there's a new update that is still 2.1
But to go to 2.2, I'd update the 'pin@2.1' comment to 'pin@2.2'

This tool felt reasonable to use when I was making the changes, and reduces the effort.

Reply:
It sure does. But again, for the unwary user it may be better to state explicitly that the comments _must_ be preserved for the tool to work again in case of an update (and that the version/reference update _must_ happen in the comment, not in the code).

What surprises me, though, is that it is possible that the hash may change upon re-running the tool: How/why can this happen? Wouldn't this be a tell-tale sign of someone tampering with the versioned/referenced action?

Reply:
Updated comments for first para.

On the second topic, agree, it would be something to look into. Normally a version won't be republished, but it could be. If the hash changes, it's different code. Any change would show up on a PR diff, and as part of a review it would be an appropriate question to ask.

without the hash mechanism in place we would silently consume the modified code in our build process... this way we at least go through a change review process.